src/qibo/ui/plot_state.py

```diff
@@ -0,0 +1,203 @@
+import matplotlib.pyplot as plt
+import numpy as np
+from matplotlib.colors import LinearSegmentedColormap, Normalize
+from mpl_toolkits.mplot3d.art3d import Poly3DCollection
+
+import qibo
+
+
+# Based on Qiskit density state plot
+# https://github.com/Qiskit/qiskit/blob/stable/2.0/qiskit/visualization/state_visualization.py#L372-L622
+def plot_density_hist(circuit, title="", alpha=0.5, colors=None):
+    """Plot the real and imaginary parts of the density matrix
+
+    Given a :class:`qibo.models.circuit.Circuit`, plots the real and imaginary parts
+    of the final density matrix as separate 3D cityscape plots, side by side, and
+    with a gray ``z=0`` plane for the imaginary part.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): Circuit to visualize.
+        title (str, optional): Title of the plot. Defaults to ``""``.
+        alpha (float, optional): Transparency level for the bars in the plot.
+            Defaults to :math:`0.5`.
+        colors (list, optional): A list of two colors for the positive and negative
+            parts of the density matrix. If ``None``, default colors will be used.
+            Defaults to ``None``.
+
+    Returns:
+        tuple: Respectively, the figure, and axes for the real and the imaginary parts.
+    """
+
+    # Execute the circuit to get the state
+    exec_circ = circuit.execute()
+
+    # if exec_circ is kind of MeasurementOutcomes, error measure gates are present
+    if isinstance(exec_circ, qibo.result.MeasurementOutcomes):
+        raise ValueError(
+            "Circuit must not contain measurement gates for density matrix visualization"
+        )
+
+    state = exec_circ.state()
+
+    # Create a density matrix from state vector
+    if not circuit.density_matrix:
+        ket = np.asarray(state)
+        state = np.outer(ket, ket.conj())
+
+    n = circuit.nqubits
+    row_names = [bin(i)[2:].zfill(n) for i in range(2**n)]
+    column_names = [bin(i)[2:].zfill(n) for i in range(2**n)]
+
+    matrix_real = state.real
+    matrix_imag = state.imag
+
+    label_y, label_x = matrix_real.shape[:2]
```

Comment:
This is calculated multiple times and it seems it could be calculated just once

Reply:
thanks a lot for spotting this!